Model/Menu/NodeRepository.php

```diff
@@ -120,6 +130,7 @@ public function getList(SearchCriteriaInterface $criteria)
         $collection->setPageSize($criteria->getPageSize());
         $objects = [];
         foreach ($collection as $objectModel) {
+            $objectModel->setHasDataChanges(false);
```

Comment:
I have a question about this line, I have tested the module, and this line doesn't make any difference (at least during my test cases), is there any reason why was it added? maybe I am missing some case and this is useful to avoid issues?

Reply:
I added this because when Magento [loads a collection](https://github.com/magento/magento2/blob/2.4.0/lib/internal/Magento/Framework/Data/Collection/AbstractDb.php#L552-L566) and [adds items data to it](https://github.com/magento/magento2/blob/2.4.0/lib/internal/Magento/Framework/Data/Collection/AbstractDb.php#L588),  
the items that get added to collection are [initially empty](https://github.com/magento/magento2/blob/2.4.0/lib/internal/Magento/Framework/Data/Collection/AbstractDb.php#L584).

When [items `addData`](https://github.com/magento/magento2/blob/2.4.0/lib/internal/Magento/Framework/DataObject.php#L44-L58) happens, and it [sets data to empty items objects](https://github.com/magento/magento2/blob/2.4.0/lib/internal/Magento/Framework/Model/AbstractModel.php#L359),  
[Magento will treat the objects as objects that have data changes](https://github.com/magento/magento2/blob/2.4.0/lib/internal/Magento/Framework/Model/AbstractModel.php#L356-L358), because `array_key_exists($key, $this->_data)`  
will be `false` as `$this->_data` is empty and getting filled with data.

Thus, when a [collection list](https://github.com/SnowdogApps/magento2-menu/blob/837ce8d2ed355bc8e19f0b61e595742b10b503db/Controller/Adminhtml/Menu/Save.php#L122) items are used to [save data](https://github.com/SnowdogApps/magento2-menu/blob/837ce8d2ed355bc8e19f0b61e595742b10b503db/Controller/Adminhtml/Menu/Save.php#L201), Magento will treat them as [modified](https://github.com/magento/magento2/blob/2.4.0/lib/internal/Magento/Framework/Model/ResourceModel/Db/AbstractDb.php#L405-L410),  
and will [update](https://github.com/magento/magento2/blob/2.4.0/lib/internal/Magento/Framework/Model/ResourceModel/Db/AbstractDb.php#L418-L420) them even without data changes.

That also means that the page cache will get unnecessarily invalidated.

I could move `$objectModel->setHasDataChanges(false);` outside the repository `getList` method and add it [here](https://github.com/SnowdogApps/magento2-menu/blob/837ce8d2ed355bc8e19f0b61e595742b10b503db/Controller/Adminhtml/Menu/Save.php#L126-L128) instead.

Reply:
I have tested it and indeed, if I remove the line, cache gets invalidated even if no changes are added to the menu. Thanks for the detailed explanation, setting this comment as resolved.